summingbird-online/src/main/scala/com/twitter/summingbird/planner/StripNamedNodes.scala

```diff
@@ -24,147 +25,67 @@ case class ProducerF[P <: Platform[P]](oldSources: List[Producer[P, Any]],
-        List(),
-        p,
-        { producerL: List[Producer[P, Any]] => p }
-      )
-
-      case p @ IdentityKeyedProducer(producer) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(producer = castToPair(producerL(0))) }
-      )
-
-      case p @ OptionMappedProducer(producer, _) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(producer = producerL(0)) }
-      )
-
-      case p @ FlatMappedProducer(producer, _) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(producer = producerL(0)) }
-      )
-
-      case p @ ValueFlatMappedProducer(producer, _) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(producer = castToPair(producerL(0))) }
-      )
-
-      case p @ KeyFlatMappedProducer(producer, _) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(producer = castToPair(producerL(0))) }
-      )
-
-      case p @ MergedProducer(oL, oR) => ProducerF(
-        List(oL, oR),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(left = producerL(0), right = producerL(1)) }
-      )
-
-      case p @ LeftJoinedProducer(producer, _) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(left = castToPair(producerL(0))) }
-      )
-
-      case p @ Summer(producer, _, _) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(producer = castToPair(producerL(0))) }
-      )
-
-      case p @ WrittenProducer(producer, _) => ProducerF(
-        List(producer),
-        p,
-        { producerL: List[Producer[P, Any]] => p.copy(producer = producerL(0)) }
-      )
-    }
-  }
-
-  def toFunctional[P <: Platform[P]](tail: Producer[P, Any]) =
-    graph
-      .dagDepth(Producer.entireGraphOf(tail))(Producer.parentsOf(_))
-      .toSeq
-      .groupBy(_._2)
-      .mapValues(_.map(_._1))
-      .mapValues(_.map(functionize(_)))
-      .toSeq
-
-  def mutateGraph[P <: Platform[P]](tail: Producer[P, Any], op: PartialFunction[Producer[P, Any], Option[Producer[P, Any]]]) = {
-    val newT: Option[Producer[P, Any]] = None
-    val x = toFunctional(tail).sortBy(_._1)
-    x.map(_._2).foldLeft((newT, Map[Producer[P, Any], Producer[P, Any]]())) {
-      case ((optLast, curMap), v) =>
-        processLevel(optLast, v, curMap, op)
+      case Source(src) => Some(src)
+      case OptionMappedProducer(_, fn) => Some(fn)
+      case FlatMappedProducer(_, fn) => Some(fn)
+      case ValueFlatMappedProducer(_, fn) => Some(fn)
+      case KeyFlatMappedProducer(_, fn) => Some(fn)
+      case LeftJoinedProducer(_, serv) => Some(serv)
+      case Summer(_, store, semi) => Some((store, semi))
+      case WrittenProducer(_, sink) => Some(sink)
+      case _ => None
```

Comment:
I'm a little concerned about this in that we must ensure this is an exhaustive list here, if any new elements of the graph get added (like the addition of the ValueFlatMappedProducer), we will have a tough time realizing its the only one that doesn't get options. If we explicitly call out all types and map the others to none scala should let us know the match isn't exhaustive if new ones are added?